File: backend/timebank/src/main/java/kookmin/software/capstone2023/timebank/presentation/api/v1/UserController.java
package kookmin.software.capstone2023.timebank.presentation.api.v1;

import kookmin.software.capstone2023.timebank.application.exception.NotFoundException;
import kookmin.software.capstone2023.timebank.application.service.account.AccountFinder;
import kookmin.software.capstone2023.timebank.application.service.auth.AccountLoginService;
import kookmin.software.capstone2023.timebank.application.service.auth.AccountRegisterService;
import kookmin.software.capstone2023.timebank.application.service.user.UserFinder;
import kookmin.software.capstone2023.timebank.application.service.user.UserUpdateService;
import kookmin.software.capstone2023.timebank.application.service.user.UserWithdrawalService;
import kookmin.software.capstone2023.timebank.domain.model.AccountType;
import kookmin.software.capstone2023.timebank.presentation.api.RequestAttributes;
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserAuthentication;
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserContext;
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.*;
import org.springframework.http.HttpStatus;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("api/v1/users")
public class UserController {

    private final AccountFinder accountFinder;
    private final AccountLoginService accountLoginService;
    private final AccountRegisterService accountRegisterService;
    private final UserFinder userFinder;
    private final UserUpdateService userUpdateService;
    private final UserWithdrawalService userWithdrawalService;

    public UserController(
            AccountFinder accountFinder,
            AccountLoginService accountLoginService,
            AccountRegisterService accountRegisterService,
            UserFinder userFinder,
            UserUpdateService userUpdateService,
            UserWithdrawalService userWithdrawalService) {
        this.accountFinder = accountFinder;
        this.accountLoginService = accountLoginService;
        this.accountRegisterService = accountRegisterService;
        this.userFinder = userFinder;
        this.userUpdateService = userUpdateService;
        this.userWithdrawalService = userWithdrawalService;
    }

    @PostMapping("login")
    public UserLoginResponseData loginUser(
            @Validated @RequestBody UserLoginRequestData data) {
        var loginData = accountLoginService.login(data.toAuthenticationRequest());

        return new UserLoginResponseData(loginData.getAccessToken());
    }

    @PostMapping("register")
    @ResponseStatus(HttpStatus.CREATED)
    public void registerUser(
            @Validated @RequestBody UserRegisterRequestData data) {
        accountRegisterService.register(
                data.toAuthenticationRequest(),
                data.getName(),
                data.getPhoneNumber(),
                data.getGender(),
                data.getBirthday(),
                AccountType.INDIVIDUAL
        );
    }

    @UserAuthentication
    @GetMapping("me")
    public CurrentUserResponseData getCurrentUser(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext) {
        var user = userFinder.findById(userContext.getUserId());
        if (user == null) {
            throw new NotFoundException("유저 정보를 찾을 수 없습니다.");
        }

        return new CurrentUserResponseData(
                user.getId(),
                user.getName(),
                user.getPhoneNumber(),
                user.getGender(),
                user.getBirthday(),
                AccountResponseData.fromDomain(user.getAccount())
        );
    }

    @UserAuthentication
    @PutMapping("me")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void updateUserInfo(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext,
            @RequestBody UserUpdateRequestData data) {
        userUpdateService.updateUserInfo(
                userContext.getUserId(),
                data.getName(),
                data.getPhoneNumber(),
                data.getGender(),
                data.getBirthday()
        );
    }

    @UserAuthentication
    @PutMapping("me/password")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void updateUserPassword(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext,
            @RequestBody UserUpdatePasswordRequestData data) {
        userUpdateService.updatePassword(
                userContext.getUserId(),
                data.getPassword()
        );
    }

    @UserAuthentication
    @DeleteMapping("me")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void withdrawalUser(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext) {
        userWithdrawalService.withdrawal(userContext.getUserId());
    }
}

File: src/main/java/kookmin/software/capstone2023/timebank/presentation/api/v1/UserController.kt
package kookmin.software.capstone2023.timebank.presentation.api.v1

import kookmin.software.capstone2023.timebank.application.exception.NotFoundException
import kookmin.software.capstone2023.timebank.application.service.account.AccountFinder
import kookmin.software.capstone2023.timebank.application.service.auth.AccountLoginService
import kookmin.software.capstone2023.timebank.application.service.auth.AccountRegisterService
import kookmin.software.capstone2023.timebank.application.service.user.UserFinder
import kookmin.software.capstone2023.timebank.application.service.user.UserUpdateService
import kookmin.software.capstone2023.timebank.application.service.user.UserWithdrawalService
import kookmin.software.capstone2023.timebank.domain.model.AccountType
import kookmin.software.capstone2023.timebank.presentation.api.RequestAttributes
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserAuthentication
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserContext
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.AccountResponseData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.CurrentUserResponseData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.UserLoginRequestData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.UserLoginResponseData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.UserRegisterRequestData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.UserUpdatePasswordRequestData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.UserUpdateRequestData
import org.springframework.http.HttpStatus
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestAttribute
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("api/v1/users")
class UserController(
    private val accountFinder: AccountFinder,
    private val accountLoginService: AccountLoginService,
    private val accountRegisterService: AccountRegisterService,
    private val userFinder: UserFinder,
    private val userUpdateService: UserUpdateService,
    private val userWithdrawalService: UserWithdrawalService,
) {
    @PostMapping("login")
    fun loginUser(
        @Validated @RequestBody
        data: UserLoginRequestData,
    ): UserLoginResponseData {
        val loginData = accountLoginService.login(data.toAuthenticationRequest())

        return UserLoginResponseData(
            accessToken = loginData.accessToken,
        )
    }

    @PostMapping("register")
    @ResponseStatus(HttpStatus.CREATED)
    fun registerUser(
        @Validated @RequestBody
        data: UserRegisterRequestData,
    ) {
        accountRegisterService.register(
            authentication = data.toAuthenticationRequest(),
            name = data.name,
            phoneNumber = data.phoneNumber,
            gender = data.gender,
            birthday = data.birthday,
            accountType = AccountType.INDIVIDUAL,
        )
    }

    @UserAuthentication
    @GetMapping("me")
    fun getCurrentUser(
        @RequestAttribute(RequestAttributes.USER_CONTEXT)
        userContext: UserContext,
    ): CurrentUserResponseData {
        val user = userFinder.findById(userContext.userId)
            ?: throw NotFoundException(message = "유저 정보를 찾을 수 없습니다.")

        return CurrentUserResponseData(
            id = user.id,
            name = user.name,
            phoneNumber = user.phoneNumber,
            gender = user.gender,
            birthday = user.birthday,
            account = AccountResponseData.fromDomain(user.account),
        )
    }

    @UserAuthentication
    @PutMapping("me")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updateUserInfo(
        @RequestAttribute(RequestAttributes.USER_CONTEXT)
        userContext: UserContext,
        @RequestBody
        data: UserUpdateRequestData,
    ) {
        userUpdateService.updateUserInfo(
            userId = userContext.userId,
            name = data.name,
            phoneNumber = data.phoneNumber,
            gender = data.gender,
            birthday = data.birthday,
        )
    }

    @UserAuthentication
    @PutMapping("me/password")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updateUserPassword(
        @RequestAttribute(RequestAttributes.USER_CONTEXT)
        userContext: UserContext,
        @RequestBody
        data: UserUpdatePasswordRequestData,
    ) {
        userUpdateService.updatePassword(
            userId = userContext.userId,
            password = data.password,
        )
    }

    @UserAuthentication
    @DeleteMapping("me")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun withdrawalUser(
        @RequestAttribute(RequestAttributes.USER_CONTEXT)
        userContext: UserContext,
    ) {
        userWithdrawalService.withdrawal(userContext.userId)
    }
}
